pkg/api/resource/typename.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"fmt"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// Typename is a generic representation for all compute resource typenames.
+// It helps enforce a specific format for naming compute resource types by providing
+// convinient input parsing methods.
+//
+// The naming format for a compute resource type is described below:
+//
+// A fully-qualified compute resource typename is constructed from a
+// DNS-style subdomain with at least one dot, a slash /, and
+// a path comprised of one or more segments separated by slashes.
+// Specifically,
+// 1. The subdomain must conform to RFC 1035 section 2.3.1 'subdomain' syntax
+//    (e.g., kubernetes.io, myveryown.org).
+// 2. The path must conform to RFC 3986 URI path-rootless syntax
+//    (e.g., memory, shinyNewResource/v2), save that it must not use dot-segments
+// 3. Typenames are treated as literal strings, and neither escaped nor case-converted.
+//    This means that case is signifcant (unlike RFC 1035 subdomains) and
+//    paths should avoid characters that need percent-encoding.
+//
+// TODO(vishh): Clarify length requirements.
+//
+// As a shorthand, any compute resource typename that does not start with a subdomain
+// and a slash will automatically be prefixed with the built-in Kubernetes namespace,
+// "kubernetes.io/resources/" in order to fully-qualify it.
+// This namespace is reserved for code in the open source Kubernetes repository.
+// As a result, all user typenames *MUST* be fully qualified, and cannot be created in this namespace.
+//
+// The recommended best practice is to use a lowercase subdomain and Go-like ASCII camelCase
+// for path components. Some example typenames include "memory"
+// (which will be fully-qualified as "kubernetes.io/resources/memory"),
+// and "myveryown.org/shinyNewResource/v2".
+//
+type Typename struct {
```

Comment:
If this is going to be exposed via the API, it needs to go in api/types.go and api/v1beta3/types.go.